### ISSUE TITLE
False positive TCH004

### ISSUE THREAD

**Issue by tlambert03** [astral-sh/ruff#10255]
I occasionally use the following pattern in an `__init__` when I want to publicly re-export a name (`"RiskyExport"`) from a sub-module (`dangerous_submodule`), without greedily importing that submodule (for example, if it requires an additional dependency that may or may not be installed)

```python
from typing import TYPE_CHECKING, Any

__all__ = ["RiskyExport"]

if TYPE_CHECKING:
    from .dangerous_submodule import RiskyExport

def __getattr__(name: str) -> Any:
    if name == "RiskyExport":
        from .dangerous_submodule import RiskyExport  # here's the actual import

        return RiskyExport
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

ruff will complain with:

- Move import `.dangerous_submodule. RiskyExport` out of type-checking block. Import is used for more than type hinting. Ruff(TCH004)

but it's *not* used for more than type hinting at the top level, because i import it again inside of `__getattr__` if someone explicitly imports that name.

Note that if I remove `RiskyExport` from `__all__`, then the error changes to F401 (imported by unused)

**Comment by MichaReiser** [comment]
That does make sense to me. To summarize what I understand you're suggesting: The idea is to change the heuristic to not flag `RiskyExport` as runtime type because it gets re-imported as runtime type in all paths referencing the runtime value. 

I don't know if we have the capabilities to do this today, because it requires that the analysis is path-sensitive (it doesn't just track whether a symbol is imported and used as typing or runtime type, but doing so for each possible path through your program)

**Comment by tlambert03** [comment]
yep, your summarized understanding is correct.  and I agree it is path sensitive, you'd need to check whether the name being used for more than type hinting wasn't available in either its global *or* local namespace
